src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -35,8 +36,9 @@
 #include "utils/snapmgr.h"
 
 
-static void ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
-									  Query *selectQuery, EState *executorState);
+static HTAB * ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
```

Comment:
We've briefly chatted with @marcocitus on this function, specifically its API.

Basically, the function becomes able to handle two very powerful/important things. Get a select query, execute it, either insert into a distributed table or split the results in a way that the results become co-located intermediate results with the target shards of a given relation.

The latter of the above could be a useful infrastructure to build more sophisticated query planning in the future.

So, the summary is that, @velioglu could you work/think on the API of this function a bit more and potentially make the distinction more explicit? 

Let's discuss the high-level changes you've in mind before pushing any changes for this.